File: app/js/tabs/diseases.js
// Diseases tab — the diseases block (may differ from interpretations).

import { hbar, barCell } from '../core/charts.js';
import { fmtYears, esc } from '../core/format.js';
import { renderTable } from '../core/tables.js';

export default {
  id: 'diseases',
  label: 'Diseases',
  computeKey: 'diseases',

  render(section, d) {
    section.innerHTML = `
      <div class="section-title">Diseases Block</div>
      <div class="section-subtitle">From the diseases field (may differ from interpretations)</div>
      <div class="card" style="margin-bottom:20px">
        <h3>Top 30 Diseases</h3>
        <div class="chart-wrap" id="chart-dis-wrap" style="height:360px"><canvas id="chart-dis"></canvas></div>
      </div>
      <div class="card">
        <h3>Full Table</h3>
        <div id="tbl-diseases"></div>
      </div>`;

    document.getElementById('chart-dis-wrap').style.height =
      Math.max(200, d.chart.labels.length * 22) + 'px';
    hbar('dis', 'chart-dis', d.chart.labels, d.chart.counts, '#60a5fa');

    const cols = [
      { key: 'label', label: 'Disease', render: r => `<span>${esc(r.label)}</span>` },
      { key: 'id', label: 'MONDO ID', render: r => `<span style="color:var(--muted);font-size:11px">${esc(r.id)}</span>` },
      { key: 'count', label: 'Patients', render: r => barCell(r.count, r.pct, 'var(--accent)') },
      { key: 'median_onset', label: 'Median Onset', render: r => fmtYears(r.median_onset) },
      { key: 'min_onset', label: 'Earliest Onset', render: r => fmtYears(r.min_onset) },
    ];
    renderTable('tbl-diseases', d.table, cols,
      { defaultSort: 'count', defaultDir: 'desc', searchKeys: ['label', 'id'] });
  },
};

File: app/py/cohort_stats/diseases.py
"""Diseases tab — the ``diseases`` block (may differ from interpretations)."""

from __future__ import annotations

from .utils import iso_to_years, median_of


def compute(packets: list) -> dict:
    n = len(packets)
    dis_map: dict = {}

    for p in packets:
        for d in p.get("diseases") or []:
            term = d.get("term") or {}
            if not term.get("id"):
                continue
            entry = dis_map.setdefault(
                term["id"],
                {"id": term["id"], "label": term.get("label") or "—", "count": 0, "onsets": []},
            )
            entry["count"] += 1
            yr = iso_to_years(((d.get("onset") or {}).get("age") or {}).get("iso8601duration"))
            if yr is not None:
                entry["onsets"].append(yr)

    rows = sorted(dis_map.values(), key=lambda d: d["count"], reverse=True)
    table = [
        {
            "id": d["id"],
            "label": d["label"],
            "count": d["count"],
            "pct": (d["count"] / n * 100) if n else 0.0,
            "median_onset": median_of(sorted(d["onsets"])),
            "min_onset": min(d["onsets"]) if d["onsets"] else None,
        }
        for d in rows
    ]
    top30 = table[:30][::-1]

    return {
        "table": table,
        "chart": {"labels": [d["label"] for d in top30], "counts": [d["count"] for d in top30]},
    }
